comment_scraper.py

```diff
@@ -1,87 +1,108 @@
-# This sample shows you how to scrape comments from a project.
-# Pass the root_asset_id and a developer token in to get_all_project_comments
-# to scrape the comments from a project. 
+###################################
+# This scraper shows you how to gather comments from
+# a Frame.io project and write to a tidy CSV.
+# Comments are gathered recursively from the 
+# Folders, files and version stacks within your project.
+###################################
 
-from frameioclient import FrameioClient
-import requests, json, csv, itertools
-
-ROOT_ASSET_ID = "Put your root asset ID here."
-TOKEN = "Put your developer token here."
-
-# Function for grabbing comments from assets. It's used by
-# get_all_project_comments. It takes an initialized client,
-# asset_id, and reference to an empty list. It returns
-# all comments on all assets.
-
-def all_comments(client, asset_id, comment_list):
-    files = client.get_asset_children(asset_id)
 
-    for asset in files:
-        if asset['type'] == "file":
-            if asset['comment_count'] > 0:
+import csv
+import os
+from itertools import chain
 
-# You can't get the asset name or the parent ID from the get_comments call, so we are saving it
-# from the get_asset_children call, so we can append it to each comment for use
-# later.
+from frameioclient import FrameioClient
 
-                asset_parent_id = asset['parent_id']
-                asset_name = asset['name']
-                comments = client.get_comments(asset['id'])
-                my_comment_list = [comment for comment in comments.results]
-                for object in my_comment_list:
-                    object.update({'parent_id':asset_parent_id})
-                    object.update({'name':asset_name})
-                comment_list.append(my_comment_list)
+class ClientNotTokenized(Exception):
+    pass
+
+
+class RootAssetIDNotFound(Exception):
+    pass
+
+
+def build_comments_list(client, asset_id, comment_list):
+    """
+    Takes an initialized client, recursively builds a list of comments
+    and returns the list. (Technically, it's a list of dicts)
+    """
+    assets = client.get_asset_children(asset_id)
+
+    for asset in assets:
+        # Recurse through folders but skip the empty ones
+        if (asset['type'] == 'folder') and (asset['item_count'] > 0):
+            build_comments_list(client, asset['id'], comment_list)
+
+        if (asset['type'] == 'file') and (asset['comment_count'] > 0):
+            comments = client.get_comments(asset['id'])
+            for comment in comments:
+                # The 'get_comments" call won't return the asset name or parent ID
+                # So we'll add them to the dictionary now. 
+                comment['parent_id'] = asset['parent_id']
```

Comment:
Comments already have a `parent_id` attribute, which points to another comment (e.g. replies will have a parent comment, as indicated by `comment.parent_id`). So, it's not a great practice to override that attribute.

It'd probably be best to do something like:

```python
comment['asset'] = { 'parent_id': asset['parent_id'], 'name': asset['name'] }
```

Or, you could leverage the `include` query API to have the entire associated asset included in each comment in the API response. That would increase the size of the request response, however.

Finally, why are we preserving this in the first place? What's the end goal here?

Reply:
👍 Ah, I overlooked that `parent_id` was already returned in the get_comments call.  

I've adapted the suggested to include asset metadata in a nested dict.  For now, it's just `asset['name']` but I could see adding more data in the future to demonstrate some corner case for a customer.

The end goals is unfortunately pretty vague, but it's to help customers who want to pull down a bunch of data on what's happening in Frame.io for a number of reasons.  The main reason I included this data was because it was in the previous version of the code -- though not working.